### DISCUSSION TITLE
Use PSRAM for big buffer and have perfect uninterrupted playback.

### DISCUSSION THREAD

**Discussion by Oppa-YA** [pschatzmann/ESP32-A2DP#247] [Ideas]
I have this idea for a long time. I was using ESP32­-WROVER-­B  &  ESP32-­WROVER­-IB modules for all of my projects and this module features an additional **8 MB SPI Pseudo static RAM (PSRAM).**
Can you please add ability to enable use of this **PSRAM** in code so it will be used as a buffer?
When i use **ESP32-A2DP** and have bad signal or I walk in other from this ESP32 receiver room sound starts chopping and loosing parts of the audio, would be cool to actually use big potential of this chip with **PSRAM**.
Do you think it's possible? Is it hard to add this option?
Here some example photos of the module and also [**Datasheet link**](https://www.espressif.com/sites/default/files/documentation/esp32-wrover-b_datasheet_en.pdf
)

Thank you!

![large_aoc388_3](https://user-images.githubusercontent.com/11776599/175657998-17cd983f-fcbb-49e6-ad21-1a896428287a.jpg)  ![Espressif_ESP32-WROVER](https://user-images.githubusercontent.com/11776599/175658108-49467ad5-84ce-41c8-8c35-955f14556fc3.jpg)

**Comment by pschatzmann** [comment]
I am not sure if this will really help with the quality but it should be pretty easy to implement by creating a subclass of BluetoothA2DPSink and overwriting the methods that allocate and release the ringbuffer. 
Here is the example code: 

 ```
/**
 * @brief BluetoothA2DPSink where the Ringbuffer is created in SPI RAM 
 * 
 */
class BluetoothA2DPSinkSPIRAM : public BluetoothA2DPSink {
    public:
        BluetoothA2DPSinkSPIRAM() = default;

        /// Allocate Ringbuffer in SPI RAM
        void bt_i2s_task_start_up(void) override {

            buffer_struct = (StaticRingbuffer_t *)heap_caps_malloc(sizeof(StaticRingbuffer_t), MALLOC_CAP_SPIRAM);
            buffer_storage = (uint8_t *)heap_caps_malloc(i2s_ringbuffer_size, MALLOC_CAP_SPIRAM);

            //Create a ring buffer with manually allocated memory
            s_ringbuf_i2s = xRingbufferCreateStatic(i2s_ringbuffer_size, RINGBUF_TYPE_BYTEBUF, buffer_storage, buffer_struct);
            if (s_ringbuf_i2s == NULL) {
                ESP_LOGE(BT_AV_TAG, "xRingbufferCreate");
                return;
            }
            xTaskCreate(ccall_i2s_task_handler, "BtI2STask", i2s_stack_size, NULL, i2s_task_priority, &s_bt_i2s_task_handle);
            ESP_LOGI(BT_AV_TAG, "BtI2STask Started");
        }

        void bt_i2s_task_shut_down(void) override {
            BluetoothA2DPSink::bt_i2s_task_shut_down();
            free(buffer_struct);
            free(buffer_storage);
   
        }
    protected:
        StaticRingbuffer_t *buffer_struct=nullptr;
        uint8_t *buffer_storage = nullptr;

};
 ```
Before calling start just use the set_i2s_ringbuffer_size() to define the Ringbuffer size..

**Comment by Oppa-YA** [comment]
@pschatzmann Thank you. It's a shame I was too much noob to implement it myself, could you please help me how to use it with "bt_music_receiver_32bit_ext" example code?
I tried to create a separate `BluetoothA2DPSinkSPIRAM.h` file with your code alongside `BluetoothA2DPSink32.h` (that was in one folder with `bt_music_receiver_32bit_ext.ino`) but it surely does not work.
I also tried to add that code to to existing file `BluetoothA2DPSink32.h` instead of creating separate file - it was also don't work. 
I was getting error  `'set_i2s_ringbuffer_size' was not declared in this scope`. 
Thank you!

![2](https://user-images.githubusercontent.com/11776599/175765660-33b2c42b-ab9c-4cc8-adcd-822e3c922076.png)

**Comment by pschatzmann** [comment]
This is a regular method on the object, so it's a2dp_sink. set_i2s_ringbuffer_size(4096);
You can add the class directly into the sketch. If you put them in a separate file do not forget the includes...

**Comment by Oppa-YA** [comment]
@pschatzmann  Thank you very much for help!
After compiling and testing can confirm that it works the same as no PSRAM buffer (I mean loosing parts of sound, skipping while having a bad signal). Maybe only new thing is it introducing tiny biping sounds when loosing and getting connection again, which i did not noticed before. Otherwise the is no additional lag while i tested it by watching video, which is I was expected because i setup a huge buffer. Not sure if it use PSRAM or not. Maybe it just made in "ESP IDF code" so it will not use bigger buffer in any circumstances.
Anyway, thank you for taking time to help!

**Comment by pschatzmann** [comment]
Did you use something like the following:
BluetoothA2DPSinkSPIRAM a2dp_sink;
a2dp_sink.set_i2s_ringbuffer_size(4000000); // 4MB PS RAM

**Comment by Oppa-YA** [comment]
I see what you mean, no, i only used default line of code 
`BluetoothA2DPSink32 a2dp_sink;` and there is no `BluetoothA2DPSinkSPIRAM a2dp_sink;`, because i did not know how to combine both of them, since they define to the same thing (sink)? I use example **(bt_music_receiver_32bit_ext)** which have that external file `BluetoothA2DPSink32.h` So maybe thats why it did not work? my code looks like this:

```
#include "BluetoothA2DPSink32.h"

class BluetoothA2DPSinkSPIRAM : public BluetoothA2DPSink {
   public:
       BluetoothA2DPSinkSPIRAM() = default;

       /// Allocate Ringbuffer in SPI RAM
       void bt_i2s_task_start_up(void) override {

           buffer_struct = (StaticRingbuffer_t *)heap_caps_malloc(sizeof(StaticRingbuffer_t), MALLOC_CAP_SPIRAM);
           buffer_storage = (uint8_t *)heap_caps_malloc(i2s_ringbuffer_size, MALLOC_CAP_SPIRAM);

           //Create a ring buffer with manually allocated memory
           s_ringbuf_i2s = xRingbufferCreateStatic(i2s_ringbuffer_size, RINGBUF_TYPE_BYTEBUF, buffer_storage, buffer_struct);
           if (s_ringbuf_i2s == NULL) {
               ESP_LOGE(BT_AV_TAG, "xRingbufferCreate");
               return;
           }
           xTaskCreate(ccall_i2s_task_handler, "BtI2STask", i2s_stack_size, NULL, i2s_task_priority, &s_bt_i2s_task_handle);
           ESP_LOGI(BT_AV_TAG, "BtI2STask Started");
       }

       void bt_i2s_task_shut_down(void) override {
           BluetoothA2DPSink::bt_i2s_task_shut_down();
           free(buffer_struct);
           free(buffer_storage);
  
       }
   protected:
       StaticRingbuffer_t *buffer_struct=nullptr;
       uint8_t *buffer_storage = nullptr;
};



BluetoothA2DPSink32 a2dp_sink; // Subclass of BluetoothA2DPSink

void setup() {
  a2dp_sink.set_swap_lr_channels(true);
  a2dp_sink.set_bits_per_sample(32);
  a2dp_sink.set_i2s_ringbuffer_size(4000000);  
  a2dp_sink.start("HiFi-32bit");
}
void loop() {
  delay(1000); // do nothing
}
```

**Comment by pschatzmann** [comment]
I suggest that you forget about BluetoothA2DPSink32. This is rather exotic and usually not needed.

Just replace 
BluetoothA2DPSink32 a2dp_sink; // Subclass of BluetoothA2DPSink
with
BluetoothA2DPSinkSPIRAM a2dp_sink; // Subclass of BluetoothA2DPSink

ps. I dont really see any advantage of   a2dp_sink.set_bits_per_sample(32); unless you have a dac which only supports 32 bits. Usually 16 would be my recommended setting and the value that a2dp is providing...

**Comment by Oppa-YA** [comment]
I tried `BluetoothA2DPSinkSPIRAM a2dp_sink;` and there is no change in the behavior, same audio drops. 
I will try to delete `a2dp_sink.set_bits_per_sample(32);` and try again.
My DAC is PCM5102 which accepts 16-, 24-, and 32-Bit Audio Data.

**Comment by Oppa-YA** [comment]
Tried this simpler variant without 32-bits. At first it seems like it work better but after flashing OLD/NEW code back and forth it seems almost the same in case of audio drops. To be sure I will try to test this receiver with this new code in more controlled area outside, maybe it will act different. As for delay I did not noticed any additional delay with 4MB PSRAM buffer enabled.

```
#include "BluetoothA2DPSink.h"

class BluetoothA2DPSinkSPIRAM : public BluetoothA2DPSink {
   public:
       BluetoothA2DPSinkSPIRAM() = default;

       /// Allocate Ringbuffer in SPI RAM
       void bt_i2s_task_start_up(void) override {

           buffer_struct = (StaticRingbuffer_t *)heap_caps_malloc(sizeof(StaticRingbuffer_t), MALLOC_CAP_SPIRAM);
           buffer_storage = (uint8_t *)heap_caps_malloc(i2s_ringbuffer_size, MALLOC_CAP_SPIRAM);

           //Create a ring buffer with manually allocated memory
           s_ringbuf_i2s = xRingbufferCreateStatic(i2s_ringbuffer_size, RINGBUF_TYPE_BYTEBUF, buffer_storage, buffer_struct);
           if (s_ringbuf_i2s == NULL) {
               ESP_LOGE(BT_AV_TAG, "xRingbufferCreate");
               return;
           }
           xTaskCreate(ccall_i2s_task_handler, "BtI2STask", i2s_stack_size, NULL, i2s_task_priority, &s_bt_i2s_task_handle);
           ESP_LOGI(BT_AV_TAG, "BtI2STask Started");
       }

       void bt_i2s_task_shut_down(void) override {
           BluetoothA2DPSink::bt_i2s_task_shut_down();
           free(buffer_struct);
           free(buffer_storage);
  
       }
   protected:
       StaticRingbuffer_t *buffer_struct=nullptr;
       uint8_t *buffer_storage = nullptr;
};



BluetoothA2DPSinkSPIRAM a2dp_sink; // Subclass of BluetoothA2DPSink

void setup() {
  a2dp_sink.set_swap_lr_channels(true);
  a2dp_sink.set_i2s_ringbuffer_size(4000000);  
  a2dp_sink.start("HiFi-32bit");
}
void loop() {
  delay(1000); // do nothing
}
```

**Comment by mitchjs** [comment]
im not sure if this is related to your audio drops, it was for mine... when sending in lots of avrcp volume messages
finding the default BT_APPT task prirorty too high... get some lower congestion messages
try a2dp_sink.set_task_priority(configMAX_PRIORITIES - 10);
the only task that prob needs it be high is the i2stask (which should be max-3)

**Reply by pschatzmann** [reply]
In this case should'nt we change the default value ?

**Comment by mitchjs** [comment]
for his case? not sure... but as per the changes Espressif suggested max-10
and it worked for me,  even with the ringbuffer max-3 was just too high
and was starving another task, max-10 works solid for me... no more audio hiccups when avrcp volume messages come in
